### PR TITLE
Make incremental mode in ngc faster

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: packages/compiler/src/aot/generated_file.ts [hunk 1]
```diff
@@ -24,6 +25,16 @@ export class GeneratedFile {
...
   }
+
+  isEquivalent(other: GeneratedFile): boolean {
+    if (this.genFileUrl !== other.genFileUrl) {
+      return false;
+    }
+    if (this.source) {
+      return this.source === other.source;
+    }
+    return o.areAllEquivalent(this.stmts !, other.stmts !);
```
Comment (chuckjaz) [inline packages/compiler/src/aot/generated_file.ts]: consider `if ((this.stmts == null)  != (other.stmts == null)) { return false; }` and prefixing the this line with `this.stmts != null`

File: packages/compiler/src/output/output_ast.ts [hunk 2]
```diff
@@ -222,6 +249,8 @@ export class ReadVarExpr extends Expression {
       this.builtin = <BuiltinVar>name;
     }
   }
+  isEquivalent(e: Expression): boolean { return e instanceof ReadVarExpr && this.name === e.name; }
```
Comment (chuckjaz) [inline packages/compiler/src/output/output_ast.ts]: `builtIn`?

File: packages/compiler/src/output/output_ast.ts [hunk 3]
```diff
@@ -261,6 +293,10 @@ export class WriteKeyExpr extends Expression {
     super(type || value.type, sourceSpan);
     this.value = value;
   }
+  isEquivalent(e: Expression): boolean {
+    return e instanceof WriteKeyExpr && this.index.isEquivalent(e.index) &&
```
Comment (chuckjaz) [inline packages/compiler/src/output/output_ast.ts]: `receiver`?

File: packages/compiler/src/output/output_ast.ts [hunk 4]
```diff
@@ -335,6 +387,9 @@ export class LiteralExpr extends Expression {
   constructor(public value: any, type?: Type|null, sourceSpan?: ParseSourceSpan|null) {
```
Comment (chuckjaz) [inline packages/compiler/src/output/output_ast.ts]: Consider changing this to `number|string|boolean|null|undefined` to make it clear the comparison below works.

**Comment by tbosch** [comment]
Also: make the number of max files before we do a full emit a top level constant

File: packages/compiler/src/output/output_ast.ts [hunk 5]
```diff
@@ -347,6 +402,10 @@ export class ExternalExpr extends Expression {
       sourceSpan?: ParseSourceSpan|null) {
     super(type, sourceSpan);
   }
+  isEquivalent(e: Expression): boolean {
+    return e instanceof ExternalExpr && this.value.name === e.value.name &&
+        this.value.moduleName === e.value.moduleName && this.value.runtime === e.value.runtime;
```
Comment (chuckjaz) [inline packages/compiler/src/output/output_ast.ts]: source `ExternalRefernce` have an `isEquivialent`?

File: packages/compiler/src/output/output_ast.ts [hunk 6]
```diff
@@ -410,6 +482,9 @@ export class FunctionExpr extends Expression {
       sourceSpan?: ParseSourceSpan|null) {
     super(type, sourceSpan);
   }
+  isEquivalent(e: Expression): boolean {
+    return e instanceof FunctionExpr && areAllEquivalent(this.statements, e.statements);
```
Comment (chuckjaz) [inline packages/compiler/src/output/output_ast.ts]: `params`?

File: packages/compiler/src/output/output_ast.ts [hunk 7]
```diff
@@ -576,6 +683,10 @@ export class DeclareFunctionStmt extends Statement {
     super(modifiers, sourceSpan);
     this.type = type || null;
   }
+  isEquivalent(stmt: Statement): boolean {
+    return stmt instanceof DeclareFunctionStmt &&
```
Comment (chuckjaz) [inline packages/compiler/src/output/output_ast.ts]: `params`?

File: packages/compiler/src/output/output_ast.ts [hunk 8]
```diff
@@ -646,6 +769,12 @@ export class ClassStmt extends Statement {
       sourceSpan?: ParseSourceSpan|null) {
     super(modifiers, sourceSpan);
   }
+  isEquivalent(stmt: Statement): boolean {
+    return stmt instanceof ClassStmt && this.name === stmt.name &&
```
Comment (chuckjaz) [inline packages/compiler/src/output/output_ast.ts]: `fields`?
`parent`?

**Review by chuckjaz — CHANGES_REQUESTED** [review]

File: packages/compiler/src/output/output_ast.ts (hunk 5 above)
Comment (tbosch) [reply packages/compiler/src/output/output_ast.ts]: No, we use it more as an interface. Changing to an actual interface.

File: packages/compiler/src/aot/generated_file.ts (hunk 1 above)
Comment (tbosch) [reply packages/compiler/src/aot/generated_file.ts]: Done

File: packages/compiler/src/output/output_ast.ts (hunk 2 above)
Comment (tbosch) [reply packages/compiler/src/output/output_ast.ts]: Done

File: packages/compiler/src/output/output_ast.ts (hunk 3 above)
Comment (tbosch) [reply packages/compiler/src/output/output_ast.ts]: Done

File: packages/compiler/src/output/output_ast.ts (hunk 4 above)
Comment (tbosch) [reply packages/compiler/src/output/output_ast.ts]: Done

File: packages/compiler/src/output/output_ast.ts (hunk 5 above)
Comment (tbosch) [reply packages/compiler/src/output/output_ast.ts]: Actually, this is used in a ton of places. Adding a comment for now that we also use it as an interface.

File: packages/compiler/src/output/output_ast.ts (hunk 6 above)
Comment (tbosch) [reply packages/compiler/src/output/output_ast.ts]: Done

File: packages/compiler/src/output/output_ast.ts (hunk 7 above)
Comment (tbosch) [reply packages/compiler/src/output/output_ast.ts]: Done

File: packages/compiler/src/output/output_ast.ts (hunk 8 above)
Comment (tbosch) [reply packages/compiler/src/output/output_ast.ts]: Done

**Review by chuckjaz — APPROVED** [review]

**Review by IgorMinar — APPROVED** [review]

**Pull request merged**